File: programovani_v_jave/cviceni_hnetynka/du/justify/WordReader.java
package com.mypackage.task4;

import java.io.BufferedReader;
import java.io.IOException;
import java.util.Scanner;

public class WordReader {

    private BufferedReader r;
    private boolean wordState = false;
    private boolean paragraphEnded = false;
    private boolean readNewLine = false;
    private StringBuilder currentWord = new StringBuilder();


    WordReader(BufferedReader r) {
        this.r = r;
    }

    public Token getNextToken() throws IOException {

        int c;
        while ((c = r.read()) != -1) {
            char ch = (char) c;

            if (ch == '\n') {
                if (readNewLine) paragraphEnded = true;
                readNewLine = true;
            }


            if (!Character.isWhitespace(ch)) {
                readNewLine = false;
                wordState = true;
                currentWord.append(ch);
                if (paragraphEnded) {
                    paragraphEnded = false;
                    return new EndOfParagraphToken();
                }
            } else if (wordState) {
                wordState = false;
                var word = currentWord.toString();
                currentWord = new StringBuilder();
                return new WordToken(word);
            }
        }

        if (wordState) {
            wordState = false;
            return new WordToken(currentWord.toString());
        }

        return new EndOfStreamToken();

    }
}

File: programovani_v_csharp/02-align/WordReader.cs
using System;
using System.Collections;
using System.Collections.Generic;
using System.IO;
using System.Text;

class WordReader
{


    public static IEnumerable<string?> readWords(TextReader tr)
    {
        bool word = false;
        var currentWord = new StringBuilder();
        Func<char, bool> isWhiteSpace = ch => ch == ' ' || ch == '\n' || ch == '\t';

        bool readNewLine = false;

        while (tr.Peek() >= 0)
        {
            var ch = (char)tr.Read();
            if (!isWhiteSpace(ch))
            {
                word = true;
                currentWord.Append(ch);
            }
            else if (word)
            {
                word = false;
                readNewLine = false;
                yield return currentWord.ToString();
                currentWord.Clear();
            }

            if(ch == '\n')
            {
                if(readNewLine)
                {
                    readNewLine = false;
                    yield return null;
                }
                else 
                    readNewLine = true;
            }
        }

        if (word) yield return currentWord.ToString();
    }
}